File: src/java/com/nvv/cache/FIFOCacheImpl.java
package com.nvv.cache;

import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

public class FIFOCacheImpl<K, V> extends AbstractCache<K, V> {

    protected Queue<CacheElement<K, V>> mCache = new ConcurrentLinkedQueue<CacheElement<K, V>>();

    public final synchronized void put(K key, V value) {

        CacheElement<K, V> element = mTable.get(key);

        if (updateCacheElement(element, key, value)) {
            return;
        }

        CacheElement<K, V> cacheElement = createCacheElement(key, value);

        if (isFull()) {
            CacheElement<K, V> elem = mCache.poll();
            mTable.remove(elem.getKey());
        }

        mCache.offer(cacheElement);
        mTable.put(key, cacheElement);
    }

    public final synchronized V get(K key) {
        CacheElement<K, V> element = mTable.get(key);
        return element != null ? element.getValue() : null;
    }

    @Override
    public void print() {
        System.out.print("{ ");
        for (CacheElement<K, V> cacheElement : mCache) {
            System.out.print("[ " + cacheElement.getKey() + " - " + cacheElement.getValue() + " ] ");
        }
        System.out.println("}");
    }

    @Override
    protected int itemsCount() {
        return mCache.size();
    }
}

File: src/com/nvv/cache/kotlin/FIFOCacheImpl.kt
package com.nvv.cache.kotlin

import java.util.concurrent.ConcurrentLinkedQueue
import java.util.Queue

public class FIFOCacheImpl<K, V> : AbstractCache<K, V>() {

    protected val mCache : Queue<CacheElement<K, V>> = ConcurrentLinkedQueue<CacheElement<K, V>>()

    public override fun put(key: K, value: V) {
        val element = mTable.get(key)

        if (updateCacheElement(element, key, value)) {
            return;
        }

        val cacheElement = createCacheElement(key, value)

        if (isFull()) {
            val elem = mCache.poll();
            mTable.remove(elem?.key);
        }

        mCache.offer(cacheElement);
        mTable.put(key, cacheElement);
    }
    public override fun print() {
        System.out.print("{ ");
        mCache.forEach() {
            System.out.print("[ " + it.key + " - " + it.value + " ] ")
        }
        System.out.println("}");
    }

    protected override fun itemsCount(): Int {
        return mCache.size();
    }
    public override fun get(key: K): V? {
        return mTable.get(key)?.value;
    }
}